common/scala/src/main/scala/whisk/core/database/DocumentHandler.scala

```diff
@@ -0,0 +1,463 @@
+  }
+
+  def computeView(ddoc: String, view: String, js: JsObject): JsObject = view match {
+    case "activations" => computeActivationView(js)
+    case _             => throw UnsupportedView(s"$ddoc/$view")
+  }
+
+  def createKey(ddoc: String, view: String, startKey: List[Any], js: JsObject): JsArray = {
+    startKey match {
+      case (ns: String) :: Nil      => JsArray(Vector(JsString(ns)))
+      case (ns: String) :: _ :: Nil => JsArray(Vector(JsString(ns), js.fields("start")))
+      case _                        => throw UnsupportedQueryKeys("$ddoc/$view -> ($startKey, $endKey)")
+    }
+  }
+
+  private def computeActivationView(js: JsObject): JsObject = {
+    val common = js.fields.filterKeys(commonFields)
+
+    val (endTime, duration) = js.getFields("end", "start") match {
+      case Seq(JsNumber(end), JsNumber(start)) if end != 0 => (JsNumber(end), JsNumber(end - start))
+      case _                                               => (JsNull, JsNull)
+    }
+
+    val statusCode = JsHelpers.getFieldPath(js, "response", "statusCode").getOrElse(JsNull)
+
+    val result = common + ("end" -> endTime) + ("duration" -> duration) + ("statusCode" -> statusCode)
+    JsObject(result.filter(_._2 != JsNull))
+  }
+
+  protected[database] def pathFilter(js: JsObject): String = {
+    val name = js.fields("name").convertTo[String]
+    annotationValue(js, "path", { v =>
+      val p = v.convertTo[String].split(PATHSEP)
+      if (p.length == 3) p(1) + PATHSEP + name else name
+    }, name)
+  }
+
+  /**
+   * Finds and transforms annotation with matching key.
+   *
+   * @param js js object having annotations array
+   * @param key annotation key
+   * @param vtr transformer function to map annotation value
+   * @param default default value to use if no matching annotation found
+   * @return annotation value matching given key
+   */
+  protected[database] def annotationValue[T](js: JsObject, key: String, vtr: JsValue => T, default: T): T = {
+    js.fields.get("annotations") match {
+      case Some(JsArray(e)) =>
+        e.view
+          .map(_.asJsObject.getFields("key", "value"))
+          .collectFirst {
+            case Seq(JsString(`key`), v: JsValue) => vtr(v) //match annotation with given key
+          }
+          .getOrElse(default)
+      case _ => default
+    }
+  }
+
+  private def dropNull(fields: JsField*) = JsObject(fields.filter(_._2 != JsNull): _*)
+}
+
+object WhisksHandler extends SimpleHandler {
+  val ROOT_NS = "rootns"
+  private val commonFields = Set("namespace", "name", "version", "publish", "annotations", "updated")
+  private val actionFields = commonFields ++ Set("limits", "exec.binary")
+  private val packageFields = commonFields ++ Set("binding")
+  private val packagePublicFields = commonFields
+  private val ruleFields = commonFields
+  private val triggerFields = commonFields
+
+  protected val supportedTables = Set(
+    "whisks.v2.1.0/actions",
```

Comment:
should these be parameterized on the view versions?

Reply:
You mean have a separate variable for version and refer to that in name? That can be done

Here I wanted to explicitly list the actual version which current implementation is coded for. If any change in version is done in CouchDB views and config in application.conf updated this hardcoding would ensure that we take a conscious update here and ensure that implementations logic is adapted for newer version